server.go

```diff
@@ -17,18 +18,41 @@ import (
 
 // The server handles creating connections and listening on various ports.
 type Server struct {
-	n              *internal.Node
-	listeners      map[string]*internal.SSHListener
-	listen_address string
+	n         *internal.Node
+	listeners map[string]*internal.SSHListener
+	// This contains the addresses of the nodes that we are currently connecting
+	// to, so we don't try to connect to them twice at the same time.
+	currently_connecting map[string]bool
```

Comment:
Overall, the main problem with relying on the connections stored in the internal node structure is that node keeps track only of things we have already connected to. As far as it's concerned, before we've established a connection, it doesn't need to know. However, the important thing here is where we're currently trying to connect. Node knows nothing about that.